src/rust/dbv2/migrations/2024-01-17-184459_coingecko/up.sql

```diff
@@ -0,0 +1,192 @@
+    SELECT ($2 * tick_size * POW(10,COALESCE(base.decimals, 0))) / lot_size
+    FROM market_registration_events
+    LEFT JOIN api.coins base
+    ON base_account_address = base."address" AND base_module_name = base.module AND base_struct_name = base.struct
+    INNER JOIN api.coins quote
+    ON quote_account_address = quote."address" AND quote_module_name = quote.module AND quote_struct_name = quote.struct
+    WHERE market_id = $1;
+$$ LANGUAGE sql;
+
+CREATE OR REPLACE FUNCTION quote_indivisible_subunits_to_integer_price(market_id numeric, price numeric) RETURNS NUMERIC IMMUTABLE AS $$
+    SELECT ($2 * POW(10,COALESCE(base.decimals, 0)) * lot_size) / tick_size
+    FROM market_registration_events
+    LEFT JOIN api.coins base
+    ON base_account_address = base."address" AND base_module_name = base.module AND base_struct_name = base.struct
+    INNER JOIN api.coins quote
+    ON quote_account_address = quote."address" AND quote_module_name = quote.module AND quote_struct_name = quote.struct
+    WHERE market_id = $1;
+$$ LANGUAGE sql;
+
+
+CREATE OR REPLACE FUNCTION size_to_base_indivisible_subunits(market_id numeric, "size" numeric) RETURNS NUMERIC IMMUTABLE AS $$
+    SELECT $2 * lot_size
+    FROM market_registration_events
+    WHERE market_id = $1;
+$$ LANGUAGE sql;
+
+
+CREATE OR REPLACE FUNCTION size_and_price_to_quote_indivisible_subunits(market_id numeric, "size" numeric, price numeric) RETURNS NUMERIC IMMUTABLE AS $$
+    SELECT $2 * $3 * tick_size
+    FROM market_registration_events
+    WHERE market_id = $1;
+$$ LANGUAGE sql;
+
+
+CREATE OR REPLACE FUNCTION get_quote_volume_divisor_for_market(numeric) RETURNS NUMERIC IMMUTABLE AS $$
+    SELECT tick_size * POW(10::numeric,decimals::numeric)
+    FROM market_registration_events AS m
+    INNER JOIN api.coins AS c
+    ON m.quote_account_address = c."address"
+    AND m.quote_module_name = c.module
+    AND m.quote_struct_name = c.struct
+    WHERE market_id = $1;
+$$ LANGUAGE sql;
+
+
+-- Create utility functions used by the views
+CREATE FUNCTION integer_price_to_quote_nominal(market_id numeric, price numeric) RETURNS NUMERIC IMMUTABLE AS $$
+    SELECT integer_price_to_quote_indivisible_subunits(market_id, price) / get_quote_volume_divisor_for_market(market_id);
+$$ LANGUAGE sql;
+
+
+CREATE FUNCTION api.get_market_liquidity (market_id numeric, depth numeric) RETURNS NUMERIC AS $$
+    WITH mid_price AS (
+        SELECT
+          (MIN(price) FILTER (WHERE direction = 'ask') + MAX(price) FILTER (WHERE direction = 'bid')) / 2 AS mid_price
+        FROM api.orders
+        WHERE orders.market_id = $1
+        AND order_status = 'open'
+    )
+    SELECT SUM(CASE
+        WHEN direction = 'bid' THEN size_and_price_to_quote_indivisible_subunits($1,remaining_size,price)
+        ELSE size_and_price_to_quote_indivisible_subunits($1,remaining_size,mid_price)
```

Comment:
On the bid size everything looks good: figure out how much collateral has actually been posted, and report that as bid liquidity

As for ask liquidity, we are trying to price out the collateral that has actually been posted, hence we need a reference price to convert base to quote

Mid price is on the right track, except none of the asks will ever fill at the mid price. At best they'll fill at the best ask price

However, for a sufficiently large spread, even the best ask price could be too high, so I'd suggest for reference price on the ask side we use the last fill price

Because the last fill price is the most honest answer to "what is current market price of the base asset"?

So for ask liquidity I say do `ELSE size_and_price_to_quote_indivisible_subunits($1,remaining_size,last_fill_price)`

Then we don't need the mid price CTE